tests/circuit_graph/test_qasm.py

```diff
@@ -0,0 +1,740 @@
+    """Conformance tests to ensure that the CircuitGraph
+    serialized QASM conforms to the QASM standard as implemented
+    by Qiskit. Note that this test class requires Qiskit and
+    pennylane-qiskit as a dependency."""
+
+    @pytest.fixture
+    def check_dependencies(self):
+        self.qiskit = pytest.importorskip("qiskit", minversion="0.14.1")
+        pl_qiskit = pytest.importorskip("pennylane_qiskit")
+
+    def test_agrees_qiskit_plugin(self, check_dependencies):
+        """Test that the QASM generated by the CircuitGraph agrees
+        with the QASM generated by the PennyLane-Qiskit plugin."""
+        dev = qml.device("qiskit.basicaer", wires=3)
+
+        @qml.qnode(dev)
+        def qnode(x):
+            qml.Hadamard(wires=0)
+            qml.RY(x[1], wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.RX(x[0], wires=1)
+            return [
+                qml.expval(qml.PauliX(0)),
+                qml.expval(qml.PauliZ(1)),
+                qml.expval(qml.Hadamard(2)),
+            ]
+
+        qnode([0.1, 0.2])
+        res = qnode.circuit.to_openqasm()
+
+        # Note: Qiskit hardcodes in pi as a QASM constant.
+        # Here, we replace it with its numerical value.
+        expected = dev._circuit.qasm().replace("pi/4", str(np.pi / 4))
+
+        assert res == expected
+
+    def test_basis_state_agrees_qiskit_plugin(self, check_dependencies):
+        """Test that the basis state prepration QASM agrees
+        with that generated by the PennyLane-Qiskit plugin. This is
+        a useful test to ensure that we are using the correct qubit
+        ordering convention."""
+        dev = qml.device("qiskit.basicaer", wires=4)
+
+        @qml.qnode(dev)
+        def qnode(state=None):
+            qml.BasisState(state, wires=[0, 1, 2, 3])
+            return qml.expval(qml.PauliZ(0))
+
+        # construct the qnode circuit
+        qnode(state=np.array([1, 0, 1, 1]))
+        res = qnode.circuit.to_openqasm()
+        expected = dev._circuit.qasm()
+
+        assert res == expected
+
+    def test_qiskit_load_generated_qasm(self, check_dependencies):
+        """Test that the QASM generated by the CircuitGraph
+        corresponds to valid QASM, that can be loaded by Qiskit."""
+        dev = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev)
+        def qnode(x):
+            qml.Hadamard(wires=0)
+            qml.RY(x[1], wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.RX(x[0], wires=1)
+            return [
+                qml.expval(qml.PauliX(0)),
+                qml.expval(qml.PauliZ(1)),
+                qml.expval(qml.Hadamard(2)),
+            ]
+
+        params = [0.1, 0.2]
+        qnode(params)
+        qasm = qnode.circuit.to_openqasm()
+        qc = self.qiskit.QuantumCircuit.from_qasm_str(qasm)
+
+        gates = [g for g, _, _ in qc.data]
+
+        for idx, g in enumerate(gates):
+            g.wires = [q.index for q in qc.data[idx][1]]

```

Comment:
How come this is needed? :thinking: might be missing something, but it seems there are no further checks for it and it's not being used for the checks

Reply:
Oh, it's being used below 🙂 I'll add a comment to make it more clear.
```suggestion
        for idx, g in enumerate(gates):
            # attach a wires attribute to each gate, containing
            # a list of wire integers it acts on, so we can assert
            # correctness below.
            g.wires = [q.index for q in qc.data[idx][1]]
```